feat(infrastructure/proxy): add readiness probes in envoy proxy deployments

psinghal20 opened the pull request:
**What type of PR is this?**
<!--
Your PR title should be descriptive, and generally start with type that contains a subsystem name with `()` if necessary 
and summary followed by a colon. format `chore/docs/feat/fix/refactor/style/test: summary`.
Examples:
* "docs: fix grammar error"
* "feat(translator): add new feature"
* "fix: fix xx bug"
* "chore: change ci & build tools etc"
-->
This PR adds readiness probes in envoy proxy deployments to prevent pods in a not-ready state to receive traffic. 

**What this PR does / why we need it**:
This allows envoy proxy configurations and in-place upgrades without traffic drops using the configured rollout strategy.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #1567 

To access the admin APIs through the readiness probe, I had to change the listener admin address to `0.0.0.0` from `127.0.0.1` to allow the envoy to listen on other network interfaces as well. Let me know if this needs to be changed.

zirain commented on internal/xds/bootstrap/bootstrap.go:
```diff
@@ -20,7 +20,7 @@ const (
 	envoyGatewayXdsServerHost = "envoy-gateway"
 	// envoyAdminAddress is the listening address of the envoy admin interface.
-	envoyAdminAddress = "127.0.0.1"
+	envoyAdminAddress = "0.0.0.0"
```
you should not change this for security reason

zirain replied:
yes, maybe use exec, I don't think we should bring thing like pilot-agent now.

arkodg commented on internal/infrastructure/kubernetes/proxy/resource.go:
```diff
@@ -154,6 +159,14 @@ func expectedProxyContainers(infra *ir.ProxyInfra, deploymentConfig *egcfgv1a1.K
+			ReadinessProbe: &corev1.Probe{
+				ProbeHandler: corev1.ProbeHandler{
+					HTTPGet: &corev1.HTTPGetAction{
+						Path: envoyReadinessPath,
```
looks like `Host` option is also available which can be set to `localhost`

zirain replied:
IIRC, that only work for hostnetwork?

psinghal20 commented:
Hi @arkodg @zirain, I have updated the PR. I have opted for a similar approach to emissary-ingress [here](https://github.com/emissary-ingress/emissary/pull/4715/files#diff-b05b9c5111c538de6527cbf67e2069e54d7dc8394fc80e1b54321966ec2b1844). 

I have added a new listener on `0.0.0.0` interface on `19001` port. Let me know if any change is required for the ready port or listener address.

arkodg reviewed (approved):
LGTM, thanks !

arkodg commented on internal/xds/bootstrap/bootstrap.yaml.tpl:
```diff
@@ -23,6 +23,33 @@ dynamic_resources:
+        address: {{ .ReadyServer.Address }}
+        port_value: {{ .ReadyServer.Port }}
+        protocol: TCP
+    filter_chains:
```
lets also add a text based access log that outputs to stdout https://github.com/envoyproxy/gateway/blob/75e7ada61fe1db9455481f2321309e099f3d8977/internal/xds/translator/listener.go#LL57C19-L57C19

arkodg replied:
as well as the per connection setting https://github.com/envoyproxy/gateway/blob/75e7ada61fe1db9455481f2321309e099f3d8977/internal/xds/translator/listener.go#L61

zirain replied:
that will be too noise

arkodg replied:
@psinghal20 can you incorporate the `PerConnectionBufferLimitBytes` setting into the config, to reduce buffer space, tia !

LanceEa commented on internal/xds/bootstrap/bootstrap.go:
```diff
@@ -28,6 +28,10 @@ const (
 	// DefaultXdsServerPort is the default listening port of the xds-server.
 	DefaultXdsServerPort = 18000
+
+	envoyReadinessAddress = "0.0.0.0"
```
@arkodg - I'm not sure it matters for this PR but does EG at this point claim to work in an IPV6 only environment?

If so then I don't believe `0.0.0.0` would work and I think its something like `::` for IPV6 (_I'm definitely not an IPv6 expert so could be wrong_).

Maybe we just need to capture it as a separate issue but thought I would call it out just in case.

zirain replied:
FYI: https://github.com/envoyproxy/gateway/issues/184

LanceEa replied:
@zirain - cool thanks for pointing out! It looks my comment is already captured so no worries from me.

LanceEa reviewed (commented):
small comment, other than that it looks good to me.

qicz reviewed (approved):
LGTM
